Cleanup  Variables API

alyst opened the pull request:
Another round of cherry-picks from #193 that cleans up the variables API, as discussed in #199:
  * `vars()`/`nvars()` methods to get the vector of SEM model variables/count of variables.
  * `observed_vars()`/`nobserved_vars()` to get the vector of observed variables ordered as in observed data (`nobserved_vars()` replaces `n_man()` method)
  * `latent_vars()`/`nlatent_vars()` to get the vector of latent variables, preserving their order in `vars()`
  * `get_data()` renamed to `samples()`, `nsamples()` is the number of data points/samples/observations in the observed data (`nsamples()` replaces `n_obs()` method)
  * for `SemMissingData` `nmeasured_vars()` is the count of observed variables with measurements (within the given data pattern)
  * interface method definitions are moved from `types.jl` to the files, where the methods of a particular type are defined (new files are created to accommodate methods for the abstract classes, e.g. `SemSpecification`)
  * these variable API methods should work not only for the SEM specification objects, like `RAMMatrices` or `ParameterTable`, but also for the types that refer to SEM specification (`SemImply`, `AbstractSem` etc)
  * old methods using ambiguous terms `get_colnames()`, or `get_n_nodes()` removed

Per #199 discussion, an alternative to *observed*/*latent* terms would be *manifest*/*latent*. Also, *samples* could be replaced with *observations* (keeping both `nobserved_vars()` and `nobservations()` may lead to confusion), and *measured*/*missing* terms could be replaced with *observed*/*missing*. I really don't have a strong preference here, so if SEM stakeholders think *manifest*/*latent* + *observed*/*missing* is a better choice, I can update the PR.
Also, I think `obs_cov`/`obs_mean` have to be updated accordingly (`observed_cov`/`observed_mean` or `manifest_cov`/`manifest_mean`).

This PR should be really the last one that does not introduce improvements or new features, but it should help to make the improvements easier.

alyst commented:
@Maximilian-Stefan-Ernst I have also cleaned up the existing unit tests and added the tests to cover the vars/params API calls

Maximilian-Stefan-Ernst commented on src/frontend/common.jl:
```diff
@@ -0,0 +1,57 @@
+"""
+    nparams(semobj)
+
+Return the number of parameters in a SEM model associated with `semboj`.
```
```suggestion
Return the number of parameters in a SEM model associated with `semobj`.

```

Maximilian-Stefan-Ernst commented on src/frontend/common.jl:
```diff
@@ -0,0 +1,57 @@
+
+See also [`vars`](@ref).
+"""
+nvars(semobj) = length(vars(semobj))
```
This function (and the next ones) do not seem to be exported - I think we should do that.

Maximilian-Stefan-Ernst commented on src/imply/abstract.jl:
```diff
@@ -0,0 +1,31 @@
+params(imply::SemImply) = params(imply.ram_matrices)
+nparams(imply::SemImply) = nparams(imply.ram_matrices)
+
+function check_acyclic(A::AbstractMatrix)
```
Some remarks:
- I think in the code (in imply/RAM/generic.jl) a different method of this function is used, and the method defined here is never actually used.
- I think the warning here should be an info, since it's totally fine to fit models without caring about having this extra bit fo performance, and having a warning might irritate new users.
- I'm not sure we should have an @info every time the matrix is lower or upper triangular, I think most users don't really care (they only might care if it is not, and for that we inform them).

alyst replied:
* Yes, putting this method here is probably a bit premature. Eventually with the introduction of *ParamsArray* type the method from `generic.jl` would be obsolete. I'll check if this implementation could be removed from the PR.
* The warning is a current behavior. Actually, for large models it has a huge impact on performance (like 5x), which reduces the model convergence from hours to minutes.
* I agree this diagnostic should be optional. The proper way would be to have a way to configure verbosity in the package, so that this message gets displayed at verbose mode. But when users are designing SEMs and try different variants, I think it would be useful to give them an option to see such diagnostic.

Maximilian-Stefan-Ernst commented on test/unit_tests/data_input_formats.jl:
```diff
@@ -18,412 +20,391 @@ dat_missing_matrix = Matrix(dat_missing)
+        end
+    end
+
+    # FIXME actually, SemObserved should not use meanstructure and always provide obs_mean()
+    # meanstructure is a part of SEM model
+    if meanstructure
+        if !isnothing(dat_mean)
+            if hasmissing
+                @test isequal(@inferred(obs_mean(observed)), dat_mean)
+            else
+                @test isequal(@inferred(obs_mean(observed)), dat_mean)
+            end
+        else
+            # FIXME if meanstructure is present, obs_mean() should provide something (currently Missing don't support it)
+            @test (@inferred(obs_mean(observed)) isa AbstractVector{Float64}) broken = true
```
Could you maybe explain these two FIXME comments to me?

alyst replied:
* First FIXME: *meanstructure* is a property of the SEM model, not of the data, but at the moment *SemObserved* accepts *meanstructure* parameter and uses it to decide whether *obs_mean()* should be initialized or not.
  I think *obs_means(::SemObserved)* should always be available (but the SEM model would consider fitting it only when SEM's own *meanstructure* flag is enabled).
  That simplifies the logic of *SemObserved* subtypes.
 * Second FIXME: at the moment *obs_means(::SemObservedMissing)* returns a vector of observed means per pattern (vector of vectors).
   The EM-based means (single vector) are not calculated at the construction time and *obs_mean(::SemObservedMssing)* does not return it.
   There are patches in my staging branch that refactor *SemObservedMissing* and EM algorithm:
     * *obs_mean()* returns a single EM-based vector of means to conform with other *SemObserved* types
     * EM-based means and covariances are calculated during *SemObservedMssing* construction
     * pattern-specific means are made available via *SemObservedMissingPattern* type.

Maximilian-Stefan-Ernst replied:
Ah tanks, that makes sense. Regarding the first FIXME, it would probably make sense to change that. Regarding the second FIXME: I think for fitting FIML models, you don't need to compute the EM covariance, it is just used in computing FABIN 3 starting values and test statistics. And the EM algorithm can be quite costly itself, and can run into convergence problems - thats why I would not compute this at construction.

Maximilian-Stefan-Ernst replied:
Ah, so you treat the EM-based cov/mean as the observed cov/mean and then do ML on them? That's an interesting use case. I think we could also either have an argument to `SemFIML` that turns on this behavior, or we could have something like `SemEM` as an additional loss function that implements this.

Maximilian-Stefan-Ernst replied:
Alright, sounds interesting - maybe this imputation scheme would also be something we could add to the package. Maybe it's best to discuss this further when the changes are in one of the next PR's ^^.

Maximilian-Stefan-Ernst commented on test/unit_tests/specification.jl:
```diff
@@ -1,22 +1,133 @@
-    partable = ParameterTable(ram_matrices)
-    @test ram_matrices == RAMMatrices(partable)
-end
+using StenoGraphs, StructuralEquationModels
```
Do we actually need to use StenoGraphs explicitly here?

Maximilian-Stefan-Ernst commented on test/Project.toml:
```diff
@@ -12,4 +12,5 @@ Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
 SafeTestsets = "1bc83da4-3b8d-516f-aca4-4fe02f6d838f"
 SparseArrays = "2f01184e-e22b-5df5-ae63-d93ebab69eaf"
 Statistics = "10745b16-79ce-11e8-11f9-7d13ad32a3b2"
+StenoGraphs = "78862bba-adae-4a83-bb4d-33c106177f81"
```
Do we need StenoGraphs as an explicit dependency for the tests?

Maximilian-Stefan-Ernst commented:
Thanks again for this great PR, and also for adding/restructuring tests. Sorry for taking so long to review. If the comments are resolved, this is ready to merge; I also agree with the terminology.

alyst commented:
@Maximilian-Stefan-Ernst Thank you for the review. I think I've addressed your comments: params, vars and observed APIs are now exported, StenoGraph explicit dependency is removed from the tests, as well as the alternative *check_acyclic()* method.
Params suggestions look like false positives (again). I wonder if the new v3 version of julia-format action fixes that.

Maximilian-Stefan-Ernst commented:
Thanks, I will update julia-format...
